lib/src/impl/realtime/channels.dart

```diff
@@ -39,6 +46,16 @@ class RealtimePlatformChannel extends PlatformObject implements spec.RealtimeCha
     return null;
   }
 
+  Map<String, dynamic> get _payload {
+    Map<String, dynamic> payload = {
```

Comment:
@zoechi this getter is used in 3 different places.

> It also looks like it should be a method instead of a getter
Interesting! Can you share more details on why?

Reply:
Because it does not return a value of a property but instead constructs a new value
every time it's called and I think a function communicates that much better.
I don't think using getters instead of no-argument functions provides much value.

Having a clear distinction between a getter/setter that mimick a field but with some additional logic like validation,
and functions that have side effects besides just updating a field (fetch, update, construct, calculate, modify, ...) makes it easier to reason about the code by just looking st the name and the kind without digging into every implementation.

Reply:
> @zoechi this getter is used in 3 different places.
> 
> > It also looks like it should be a method instead of a getter
> 
> Interesting! Can you share more details on why?

Ok, I missed that. Then I'd keep it and just change it into a function.

The final decision if such style rules should be followed lies not by me of course.
We can wait for others to chime in first.

Reply:
Right, that is a valid point!

But, when I look at this from the caller perspective, a getter feels more intuitive than a method 😉 

Meanwhile, I've optimized with suggested syntax at 3958791aeee753103087bb8f8c94ca37fd2f2bd4

Reply:
I think I know what you mean with "more intuitive" here, 
but when reading the code later (caller site) it reveals less about what the code does.

It is just a convention, but if used consistently I think it can make it easier to read the code.
It's always helpful for understanding from a short glance whether the code does some actual work or only returns a value that is already there.

For about convenience when writing see also https://www.goodreads.com/quotes/835238-indeed-the-ratio-of-time-spent-reading-versus-writing-is

Reply:
@QuintinWillison do you have an opinion on this?

Reply:
I'm honestly on the fence with this one. I can see the point of view from both sides.

Assuming I'm reading the Dart correctly, are `name` and `options` fixed (immutable) for the life of this `RealtimePlatformChannel` instance? In which case perhaps a property get style is more intuitive.

But if that's not the case then I would argue a new method name like `_createPayload()` would be more descriptive.

Reply:
In this case it probably should be something like

```dart
Map<String,dynamic> __payload;
Map<String, dynamic> get _payload => __payload ??=
    {
      "channel": name,
      if(options != null) "options": options
    };
  }
```
to avoid creating a new instance every time it is read.
Then I'd also agree on a getter.

It this case it might also be a field initialized in the constructor, but above approach with field and getter delays creation until it's actually needed which seems to be the intention here.

Reply:
Is the double underscore prefix a convention for lazily computed getters in Dart? 🤔

Reply:
No I'm not aware of such an official convention, but usually a backing field for a public getter has a leading underscore to make it private.
If the getter is private already, this would cause a conflict, so I add a 2nd underscore.
Not pretty, but I had expected this to be intuitive ;-)
So, it's just my personal convention.

Reply:
@QuintinWillison I admit this is a bit of an edge case, but I need a few such discussions to figure out what rules the code should follow. 

What really tripped me off was this code
https://github.com/ably/ably-flutter/blob/0c2f7074b7ffe6abcb5ba0a0a99aed7ab4917bd9/lib/src/impl/platform_object.dart#L18

which could even be a field access from the look of it which would be a no-op.

I found the distinction between method and getter/setter usually helpful, 
but I have seen even people on the Dart core team argue about not promoting it as official style guide rule.

If we do not make a distinction between 0-argument method and getter
and just use a getter everywhere if no parameters are required 
then I think we give up some expressiveness of the language.
Then one could argue for the sake of code consistency to ban methods without parameters completely and consistently make them getters. Which I would find quite ugly.
I also don't see a benefit worth mentioning from omitting `()`. 

If we want to keep using 0-argument functions, then I think *some* rule when to use which is helpful.
The above code example with the `handle` reminded me, that the distinction between "property" to access existing state and have no side effects, and "functions" who are used to do calculations and/or are called only for its side effects is actually useful. (functions also should have verb names like `calc...`, `fetch...`)

I'm also fine with getters like `fullName` who concatenate strings, or `grossPrice` from net price and tax,
especially when the results are value types which are not compared by identity. 

In this current case it might not be relevant that a new map is created with every access.
In Angular for example where object identity is used for change detection, this would cause an endless loop that hangs the app, because every time it checks if the model stablilized it would result in `false` because it gets a different instance to compare every time.

If you say we do not want such rules and everyone can write code like he wants, then I can accept that as well (or something in between)

Reply:
I agree that code looks odd. And the comment doesn't really help! This probably can be improved. @tiholic, what do you think? Can you explain what this code is doing?

The `calc...` and `fetch...` method naming conventions you suggest, @zoechi, are what I would choose as 0-argument functions do have their place in my view. They remind me of Apple's Cocoa naming conventions (e.g. `create` prefix on class methods).

Reply:
>  `this.handle;  //fetching asynchronous handle proactively... `

@zoechi you pointed out the right snippet. Even I was very unhappy about that line of code, but taking into consideration that the `handle` becomes a private property once fetched from platform. 

I decided it better be a getter by weighing its usage:
- creating instance on platform: once in lifetime of that Rest/Realtime instance _(a method will be intuitive here)_
- requiring handle: every other time, like publish, event listeners, etc.  _(a getter is more intuitive here)_

and also considering that this is something internal/private to the SDK and will be re-visited very rarely. I added a comment trying to explain what that code is doing as it was not intuitive/self-explanatory.

@QuintinWillison After moving from async Rest/Realtime instance creation API to synchronous one, we will be creating these instances only on demand (when connecting to a channel, publishing, or starting an event listener).
So `this.handle` is **called** (well, not actually called, but _the getter is invoked_) to pro-actively create a rest instance and acquire the handle even before user code creates a channel/attaches an event listener.

Overall, I generally am more inclined towards intuitiveness of the code, but not quite sure if we should give more weightage to that here over what is right technically in all scenarios. I'd value @zoechi 's opinion as final on this aspect.

Reply:
This is not directly related to this PR but because it is related to the discussion about style and how to make code intuitive I post it here anyway.

`handle` is a getter and has the side effect to set `this._handle` and it is even used in the constructor for the side effect only.

This might be intuitive to someone who only cares about getting the `handle` from the getter,
but if something does not work as expected then even mere package users need to look and perhaps even debug into the implementation and start reasoning about what the code does, 
and then it becomes in my opinion the exact opposite of intuitive.
Not to talk about future package maintainers.

My suggestion would be to split the getter into a getter and a function.
Each following the guideline as closely as possible.
The getter is used to return the state, but if the state is not initialized then it calls the initializer.

This is still a getter with a side effect, but this kind of lazy initialization is quite common
and probably understood by most somewhat experienced Dart users
and a brief look reveals this intention. 
Especially the `??=` operator which is mostly used for exactly this use case makes this quite obvious.

In the constructor we can call the function to avoid the confusing style of calling a getter
only for its side effect.

Because the function can simplified a lot, it looks quite different than the original code,


My suggestion would be

```dart
abstract class PlatformObject {
  static const acquireHandleTimeout = const Duration(seconds: 2);

  PlatformObject() {
     _handle = _acquireHandle();
  }

  Future<int> _handle;
  Future<int> get handle => _handle;

  Future<int> _acquireHandle() => createPlatformInstance().timeout(acquireHandleTimeout);

  authUpdateComplete() 
  ...
}
```
or if it should retry in the case of a timeout the next time `handle` is accessed

```dart
abstract class PlatformObject {
  static const acquireHandleTimeout = const Duration(seconds: 2);

  PlatformObject() {
     _handle = _acquireHandle();
  }

  Future<int> _handle;
  Future<int> get handle => _handle ??= _acquireHandle();

  Future<int> _acquireHandle() => createPlatformInstance().timeout(
    acquireHandleTimeout, onTimeout: () {
      _handle = null;
      throw new TimeoutException('Acquiring handle timed out.', acquireHandleTimeout);
    });

  authUpdateComplete() 
  ...
}
```

Reply:
I like that approach. Thanks, @zoechi. We need to always be careful with "side effects". We need stuff to be obvious and easy to reason about for maintenance purposes. @tiholic can you make that change please? I would really like to resolve this thread and get this PR merged! 😓

Reply:
From the Slack discussion. To have the resolved handle value available for `toString()` we could change it to
```dart
Future<int> _acquireHandle() => createPlatformInstance().timeout(acquireHandleTimeout).then((h) => _handleValue = h);
```
and use the additional field `_handleValue` in `String toString() => '...';

Reply:
addressed both cases
payload getter: 4a82e39d2669d22f8e45f42afea49f3c8267ed76
library registration and handle acquisition flow: 29760170d7187d8d3ca082265edd67b3dca2b427